init

Artificial-Friend opened the pull request.

Artificial-Friend committed:
init

Artificial-Friend committed:
deployed to Heroku

Artificial-Friend committed:
fix

Artificial-Friend committed:
AuthenticationFilter refactored

Artificial-Friend committed:
Why do I do it for free?
login page now is fancy

Artificial-Friend commented:
Hello, my apologies for being so devoted to this task, I overdone it a lot, just ignore login.css and for your convenience you can look onto deployed result.

Max-Pochepets commented on src/main/java/core/basesyntax/controller/LoginController.java:
```diff
@@ -0,0 +1,42 @@
+        } catch (AuthenticationException e) {
+            req.setAttribute("errorMsg", e.getMessage());
+            req.getRequestDispatcher("WEB-INF/views/login.jsp").forward(req, resp);
+            return;
```
```suggestion
```

Artificial-Friend replied:
After some testing of errors I found out that you are wrong.
https://stackoverflow.com/questions/2123514/java-lang-illegalstateexception-cannot-forward-sendredirect-create-session

Artificial-Friend replied:
Partially wrong to be precise. Redirect after this return will be failed in filter but it is redundant operation since before return we can forward user straight to login page without unnecessary checking user in filter.

Max-Pochepets commented on src/main/java/core/basesyntax/model/Driver.java:
```diff
@@ -4,10 +4,32 @@
+
+    public Driver setPassword(String password) {
+        this.password = password;
+        return this;
```
if we'll get to compare two drivers, what will we get?

Max-Pochepets commented on src/main/java/core/basesyntax/security/AuthenticationServiceImpl.java:
```diff
@@ -0,0 +1,24 @@
+        if (userFromDB.getPassword().equals(password)) {
+            return userFromDB;
+        }
+        throw new AuthenticationException("Incorrect password");
```
I don't think we should tell the user what exactly did he get wrong

Max-Pochepets commented on src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,46 @@
+            return;
+        }
+        Long id = (Long) req.getSession().getAttribute(DRIVER_ID);
+        if (id == null) {
```
what if logged in user deletes himself?

Max-Pochepets commented on src/main/webapp/WEB-INF/views/driver/add.jsp:
```diff
@@ -2,16 +2,32 @@
+        Driver's license number: <input required type="text" name="licenseNumber">
+        <br>
+        Driver's login: <input required type="text" name="login">
+        Driver's password: <input required type="password" name="password">
```
what if my input will be blank?

Artificial-Friend committed:
security:
User no longer able to access site after suicide;
Invalid login error changed to ambiguous
fixed:
equals(), hashcode() add to Driver.java
removed:
login.css moved to main branch for PR readability

Artificial-Friend committed:
checkstyle

Artificial-Friend committed:
Merge branch 'main' of github.com:Artificial-Friend/jv-dao-annotation into work.09

Max-Pochepets commented on src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java:
```diff
@@ -34,6 +35,14 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
+            resp.sendRedirect("/login");
+            return;
+        }
+        try {
+            driverService.get(id);
+        } catch (NoSuchElementException e) {
+            req.setAttribute("errorMsg", "Access denied! Log in or register to get access");
```
good. but as for me, it's better not to catch runtime exceptions, but to properly process Optional in service: we can invoke `get` on optional with a check on `Presence` and if no value present - return null. I understand that returning null is a bad practice, but it's better not to show to user stacktrace of an exception as well as not to catch runtime exceptions.

Artificial-Friend replied:
I will wait for mentors answer about this.
All in all I agree with you but: 
a) Yes, I can do in service .orElseGet(null) and meaning of Optional will be only in cautioning user that there could be null.
b) When I started to changing this service I got problem that I need to change everything to return Optional and do the same thing as we done in .findByLogin(). It is a right handling of null since I don't give null or exception, I give Optional for user to decide what they want to do with it. But my PR will look like garbage if I do it, so limiting myself with task requirements I left it as is and managed to work with what I have from perspective if I am user of this service.

Artificial-Friend committed:
added memes

Artificial-Friend committed:
Merge branch 'main' of github.com:Artificial-Friend/jv-dao-annotation into work.09

Artificial-Friend committed:
Во имя отца, сына и святого духа

RomanDubovskyi commented on src/main/java/core/basesyntax/dao/jdbc/DaoJdbcDriverImpl.java:
```diff
@@ -55,8 +58,8 @@ public Driver create(Driver item) {
     public List<Driver> getAll() {
-        String query = "SELECT id, name, license_number FROM drivers WHERE deleted = false "
-                + "ORDER BY id";
+        String query = "SELECT id, name, license_number, login, password FROM drivers "
```
```suggestion
        String query = "SELECT id, name, license_number, login FROM drivers "
```
ideally, it's better not to return a users' password from back end

RomanDubovskyi commented on src/main/java/core/basesyntax/security/AuthenticationServiceImpl.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public Driver login(String login, String password) throws AuthenticationException {
+        Driver userFromDB = driverService.findByLogin(login).orElseThrow(
```
wrote about it the chat

RomanDubovskyi commented on src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,55 @@
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+        String url = req.getServletPath();
+        if (url.equals("/login") || url.equals("/drivers/add")) {
```
wrote about it in the chat

Artificial-Friend committed:
refactored: security, optimization

Artificial-Friend committed:
checkstyle

RomanDubovskyi commented on src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,60 @@
+    private static final Injector injector = Injector.getInstance("core.basesyntax");
+    private final DriverService driverService
+            = (DriverService) injector.getInstance(DriverService.class);
+    private final Set<String> links = new HashSet<>();
```
```suggestion
    private final Set<String> allowedUrls = new HashSet<>();
```
better to use more specific naming

Artificial-Friend replied:
You're right, I already forgot that those are only allowed links.

RomanDubovskyi reviewed (approved):
improve sets' naming, rest is good👍

Artificial-Friend committed:
Update src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java

AuthenticationFilter.java refactored

Co-authored-by: RomanDubovskyi <58385506+RomanDubovskyi@users.noreply.github.com>

Artificial-Friend commented on src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+        links.add("/login");
```
```suggestion
        allowedUrls.add("/login");
```

Artificial-Friend commented on src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public void init(FilterConfig filterConfig) {
+        links.add("/login");
+        links.add("/drivers/add");
```
```suggestion
        allowedUrls.add("/drivers/add");
```

Artificial-Friend commented on src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,60 @@
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+        String url = req.getServletPath();
+        if (links.contains(url)) {
```
```suggestion
        if (allowedUrls.contains(url)) {
```

Artificial-Friend committed:
Update src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java

Artificial-Friend committed:
Update src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java

Artificial-Friend committed:
Update src/main/java/core/basesyntax/web/filter/AuthenticationFilter.java

Artificial-Friend committed:
Merge branch 'main' into work.09
